qiita_db/software.py

```diff
@@ -537,29 +609,214 @@ def command(self):
-            table_cols.remove(['parameters_id', 'bigint'])
-            table_cols.remove(['param_set_name', 'character varying'])
+            cmd_reqd_params = command.required_parameters
+            cmd_opt_params = command.optional_parameters
+
+            values = {}
+            for key in cmd_reqd_params:
+                try:
+                    values[key] = parameters.pop(key)
+                except KeyError:
+                    raise qdb.exceptions.QiitaDBError(
+                        "%s. Missing required parameter: %s"
+                        % (error_msg, key))
+
+            for key in cmd_opt_params:
+                try:
+                    values[key] = parameters.pop(key)
+                except KeyError:
+                    raise qdb.exceptions.QiitaDBError(
+                        "%s. Missing optional parameter: %s"
+                        % (error_msg, key))
+
+            if parameters:
+                raise qdb.exceptions.QiitaDBError(
+                    "%s. Extra parameters: %s"
+                    % (error_msg, ', '.join(parameters.keys())))
 
-            values = self.values
+            return cls(values, command)
+
+    @classmethod
+    def from_default_params(cls, dflt_params, req_params, opt_params=None):
+        """Creates the parameter set from a `dflt_params` set
+
+        Parameters
+        ----------
+        dflt_params : qiita_db.software.DefaultParameters
+            The DefaultParameters object in which this instance is based on
+        req_params : dict of {str: object}
+            The required parameters values, keyed by parameter name
+        opt_params : dict of {str: object}, optional
+            The optional parameters to change from the default set, keyed by
+            parameter name. Default: None, use the values in `dflt_params`
 
-            result = []
-            for p_name, p_type in sorted(table_cols):
-                if p_type == 'boolean':
-                    if values[p_name]:
-                        result.append("--%s" % p_name)
-                else:
-                    result.append("--%s '%s'" % (p_name, values[p_name]))
+        Raises
+        ------
+        QiitaDBError
+            - If there are missing requried parameters
+            - If there is an unknown required ot optional parameter
+        """
+        with qdb.sql_connection.TRN:
+            command = dflt_params.command
+            cmd_req_params = command.required_parameters
+            cmd_opt_params = command.optional_parameters
+
+            missing_reqd = set(cmd_req_params) - set(req_params)
+            extra_reqd = set(req_params) - set(cmd_req_params)
+            if missing_reqd or extra_reqd:
+                raise qdb.exceptions.QiitaDBError(
+                    "Provided required parameters not expected.\n"
+                    "Missing required parameters: %s\n"
+                    "Extra required paramters: %s\n"
```

Comment:
You know, Vim has built-in spell checker ... just saying :P

On (Nov-24-15|12:03), Jose Navas wrote:
> > +        QiitaDBError
> > +            - If there are missing requried parameters
> > +            - If there is an unknown required ot optional parameter
> > +        """
> > +        with qdb.sql_connection.TRN:
> > +            command = dflt_params.command
> > +            cmd_req_params = command.required_parameters
> > +            cmd_opt_params = command.optional_parameters
> > +
> > +            missing_reqd = set(cmd_req_params) - set(req_params)
> > +            extra_reqd = set(req_params) - set(cmd_req_params)
> > +            if missing_reqd or extra_reqd:
> > +                raise qdb.exceptions.QiitaDBError(
> > +                    "Provided required parameters not expected.\n"
> > +                    "Missing required parameters: %s\n"
> > +                    "Extra required paramters: %s\n"
> 
> I really apologize for this typos - it just needs to happen once and the editor autocomplete breaks everything... did a search/repleace
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/qiita/pull/1544/files#r45786927